Add playlist drag and drop support to the classic skin

oreillymj opened the pull request:
I still use the classic skin and the only thing I wish it had was drag and drop re-arranging of playlist items.

This implements a handler on each item and saves the changed playlist in the same fashion as the up/down arrows do.
It also preserves the odd/even zebra stripes.

oreillymj committed:
Add playlist drag and drop support to the classic skin

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

michaelherger commented on HTML/Default/skin.css:
```diff
@@ -243,3 +243,14 @@ td.sort-asc .x-grid3-hd-inner, td.x-grid3-hd-menu-open .x-grid3-hd-inner {
 .sort-desc .x-grid3-sort-icon {
 	background-image: url('data:image/gif;base64,R0lGODlhDQAFAJEAAAAAAIuLi+zs7AAAACH5BAkKAAMALAAAAAANAAUAAAIPVI4hw2y5WntKWhOtzLYAADs=');
 }
+
+
+.dragging {
+    opacity: 0.4;
+}
+.drop-target-above {
+    border-top: 2px solid #ccc;
+}
+.drop-target-below {
+    border-bottom: 2px solid #ccc;
+}
```
Isn't this the wrong stylesheet? Default should not be loaded in Classic.

oreillymj replied:
There is no skin.css in Classic even though the frame refers to it.
<link rel="stylesheet" type="text/css" href="/Classic/skin.css?r=git-242791008">

~/projects/slimserver/HTML$ find . -iname skin.css
./Default/skin.css
./EN/skin.css
./Logic/skin.css

So I assumed it used the file in EN.

michaelherger replied:
Yes, the one in EN. But you modified the one in Default 😉.

michaelherger commented on HTML/EN/html/dragdrop.js:
```diff
@@ -0,0 +1,143 @@
+}
+
+
+// Get the player ID from the DOM and decode it
+function getPlayerId() {
```
The player ID should be available as the global variable `playerid` (see var.js).

oreillymj replied:
Thanks, I'll remove my implementation and use the existing variable.

michaelherger commented on HTML/EN/html/dragdrop.js:
```diff
@@ -0,0 +1,143 @@
+}
+
+
+function movePlaylistItemXHR(playerID, fromIndex, toIndex) {
+    const xhr = new XMLHttpRequest();
+    const url = `/status.html?p0=playlist&p1=move&p2=${fromIndex}&p3=${toIndex}&player=${playerID}&ajax=1`;
+    console.log("Calling URL : " + url)
+
+    xhr.onreadystatechange = function() {
+        if (xhr.readyState === 4) {
+            if (xhr.status === 200) {
+                console.log('Player ' + playerID + ' playlist item moved successfully from ' + fromIndex + ' to ' + toIndex);
+            } else {
+                console.error('Failed to move playlist item');
+            }
+        }
+    };
+    xhr.open('GET', url, true);
+    xhr.send();
+}
```
I'm pretty sure there's some server call helper function around already. If already you're re-implementing it here, I'd recommend you use the new [Fetch API](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API). 

And I'd definitely avoid using the `status.html` method. Use the `jsonrpc.js` handler instead. It allows you to use nicely formatted JSON objects for the request, and returns the same. See https://lyrion.org/reference/cli/using-the-cli/#jsonrpcjs.

oreillymj replied:
I couldn't find an update_playlist() helper function anywhere and wanted to use the method that is used by the up/down button. But I'll re-submit using fetch if that is more robust.

oreillymj committed:
Re-implemented using Fetch and removed unnecessary playerid variable logic
Added Classic/skin.css

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

oreillymj committed:
Clean up skin.css. The file in EN was empty

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

michaelherger commented on HTML/Classic/playlist.html:
```diff
@@ -6,10 +6,15 @@
 	[% PROCESS standardheader.html %]
 	<script TYPE="text/javascript">[% PROCESS html/vars.js %]</script>
 	<script TYPE="text/javascript" src="[% webroot %]html/common.js?r=[% revision %]"></script>
+        <script TYPE="text/javascript" src="[% webroot %]html/dragdrop.js?r=[% revision %]"></script>
+        <script>
+           document.addEventListener("DOMContentLoaded", function() {
+             initPlaylistDragDrop(".browsedbList");
+           });
+        </script>
```
Please use tabs rather than spaces for the indentation.

michaelherger reviewed (changes requested):
Thanks for this addition! I'll try to test it soon. It's looking good so far. Except for the whitespace: please always use tabs rather than spaces. It's how the project started decades ago 😁.

michaelherger commented on HTML/EN/html/dragdrop.js:
```diff
@@ -0,0 +1,165 @@
+    };
+    xhr.open('GET', url, true);
+    xhr.send();
+}
```
Sorry for being a pain... please remove unused code. Thanks!

oreillymj committed:
Clean of old functions and additional snaity checks. Tested on Firefox 141 and Edge 139

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

oreillymj committed:
Styling tweaks

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

oreillymj committed:
Prevent race condition seen in testing

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

oreillymj committed:
Signed-off-by: Michael OReilly <oreillymj@gmail.com>

Reload the playlist frame after a drop event to rebuild the playback control urls.

After a drop event, the play, up, down and delete icons at the right side of the playlist frame have urls which are out of synch with the frame content.
Clicking on an item without reloading the frame results in unpredictable playback etc.
The easiest fix is to reload the frame, although you could re-write the urls from each item between the moved items.
The frame normally reloads after 30s anyway, so this is a bit of a edge case in terms of behaviour.

michaelherger reviewed (commented):
Hi @oreillymj - I think this is looking good - except for the use of spaces instead of tabs for the indentation. Could you please fix that? I'd then merge it. Thanks!

oreillymj committed:
Renamed js file to avoid confusion with existing dragdrop.js. Changed Spaces to Tabs in playlist.html. Changed from page refresh to href link re-writes for affected navigation items in the playlist

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

oreillymj commented:
Hopefully the HTML is correctly formatted, but its hard to tell in nano. I've renamed the .js as there is an existing dragdrop.js and I wanted to avoid future confusion around the files purpose. I've also switched from refreshing the frame after a drag-drop event to fixing the links for the items affected.

In my testing this works, but I've left the page reload function in case I need to revert.

michaelherger reviewed (changes requested):
We're getting to the home run! The modified files look good now - thanks for the patience!

As this only applies to the Classic skin, please move both the JS and the `skin.css` file to Classic/html. And I'd prefer the old `dragndrop.js` name. "classic" doesn't need to be in the file name when it's in that folder already.

And please format those with tabs, too.

oreillymj committed:
Cleanup as requested, and spaces converted to tabs. Ran JS through https://beautifier.io/ with Indent set to TABS

Signed-off-by: Michael OReilly <oreillymj@gmail.com>

michaelherger reviewed (approved):
Thanks a lot!

oreillymj commented:
I've added a poll on the forum to see if I'm the only person using the Classic skin. Would you mind pinning it to the top of the General Discussion forum until Sept 9th, when the poll closes. Here's the link - https://forums.lyrion.org/forum/user-forums/general-discussion/1783239-poll-what-ui-skin-do-you-use#post1783239
